test/distributed/org/apache/cassandra/distributed/test/SecondaryIndexTest.java

```diff
@@ -78,31 +80,45 @@ public void after()
-            List<InetAddress> scanning = Arrays.stream(traces)
-                                               .filter(t -> t[1].toString().matches("Index mean cardinalities are v_index:[0-9]+. Scanning with v_index."))
-                                               .map(t -> (InetAddress) t[0])
-                                               .distinct().collect(Collectors.toList());
-
-            List<InetAddress> executing = Arrays.stream(traces)
-                                                .filter(t -> t[1].toString().equals("Executing read on " + tableName + " using index v_index"))
-                                                .map(t -> (InetAddress) t[0])
-                                                .distinct().collect(Collectors.toList());
-
-            Assert.assertEquals(Collections.singletonList(cluster.get(1).broadcastAddress().getAddress()), scanning);
-            Assert.assertEquals(3, executing.size());
+            Assert.assertEquals("Failed on iteration " + i, 3, result.length);
+
+            Awaitility.await("For all events in the tracing session to persist")
+                    .pollInterval(100, TimeUnit.MILLISECONDS)
+                    .atMost(10, TimeUnit.SECONDS)
+                    .untilAsserted(() -> 
+                                   {
+                                       Object[][] traces = cluster.coordinator(1)
+                                                                  .execute("SELECT source, activity FROM system_traces.events WHERE session_id = ?", 
+                                                                           ConsistencyLevel.ALL, trace);
+
+                                       List<InetAddress> scanning =
+                                               Arrays.stream(traces)
+                                                     .filter(t -> indexScanningPattern.matcher(t[1].toString()).matches())
+                                                     .map(t -> (InetAddress) t[0])
+                                                     .distinct().collect(Collectors.toList());
+
+                                       List<InetAddress> executing =
+                                               Arrays.stream(traces)
+                                                     .filter(t -> t[1].toString().equals(String.format("Executing read on " + tableName + " using index v_index_%d", seq.get())))
+                                                     .map(t -> (InetAddress) t[0])
+                                                     .distinct().collect(Collectors.toList());
+
+                                       Assert.assertEquals(Collections.singletonList(cluster.get(1).broadcastAddress().getAddress()), scanning);
+                                       Assert.assertEquals(3, executing.size());
+                                   });
```

Comment:
It might also have been possible to make some changes to the actual Tracing implementation to force pending events to persist, but it seemed cleaner to just keep changes isolated to the test.